KAFKA-15277: Design & implement support for internal Consumer delegates

kirktrue opened the pull request:
The consumer refactoring project introduced another `Consumer` implementation, creating two different, coexisting implementations of the `Consumer` interface:

* `KafkaConsumer` (AKA "existing", "legacy" consumer)
* `PrototypeAsyncConsumer` (AKA "new", "refactored" consumer)

The goal of this task is to refactor the code via the delegation pattern so that we can keep a top-level `KafkaConsumer` but then delegate to another implementation under the covers. There will be two delegates at first:

* `LegacyKafkaConsumer`
* `AsyncKafkaConsumer`

`LegacyKafkaConsumer` is essentially a renamed `KafkaConsumer`. That implementation handles the existing group protocol. `AsyncKafkaConsumer` is renamed from `PrototypeAsyncConsumer` and will implement the new consumer group protocol from KIP-848. Both of those implementations will live in the `internals` sub-package to discourage their use.

This task is part of the work to implement support for the new KIP-848 consumer group protocol.

kirktrue committed:
KAFKA-15277: Design & implement support for internal Consumer delegates

The consumer refactoring project introduced another Consumer implementation, creating two different, coexisting implementations of the Consumer interface:

* KafkaConsumer (AKA "existing", "legacy" consumer)
* PrototypeAsyncConsumer (AKA "new", "refactored" consumer)

The goal of this task is to refactor the code via the delegation pattern so that we can keep a top-level KafkaConsumer but then delegate to another implementation under the covers. There will be two delegates at first:

* LegacyKafkaConsumer
* AsyncKafkaConsumer

LegacyKafkaConsumer essentially be a renamed KafkaConsumer. That implementation handles the existing group protocol. AsyncKafkaConsumer is renamed from PrototypeAsyncConsumer and will implement the new consumer group protocol from KIP-848. Both of those implementations will live in the "internals" sub-package to discourage their use.

This task is part of the work to implement support for the new KIP-848 consumer group protocol.

kirktrue committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-15277-consumer-delegates-take-2

kirktrue committed:
Jumping through hoops to make SpotBugs happy :(

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -321,20 +320,6 @@ public class ConsumerConfig extends AbstractConfig {
      */
     static final String LEAVE_GROUP_ON_CLOSE_CONFIG = "internal.leave.group.on.close";
 
-    /**
-     * <code>internal.throw.on.fetch.stable.offset.unsupported</code>
-     * Whether or not the consumer should throw when the new stable offset feature is supported.
-     * If set to <code>true</code> then the client shall crash upon hitting it.
-     * The purpose of this flag is to prevent unexpected broker downgrade which makes
-     * the offset fetch protection against pending commit invalid. The safest approach
-     * is to fail fast to avoid introducing correctness issue.
-     *
-     * <p>
-     * Note: this is an internal configuration and could be changed in the future in a backward incompatible way
-     *
-     */
-    static final String THROW_ON_FETCH_STABLE_OFFSET_UNSUPPORTED = "internal.throw.on.fetch.stable.offset.unsupported";
-
```
I moved this to `ConsumerUtils` because `LegacyKafkaConsumer` and `AsyncKafkaConsumer` couldn't access it via package-level visibility since they're in the `internals` sub-package.

@dajac—will this be considered a breaking change? I assumed not because `THROW_ON_FETCH_STABLE_OFFSET_UNSUPPORTED` isn't public.

dajac replied:
If it is not public, it is fine.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -666,19 +651,6 @@ else if (newConfigs.get(VALUE_DESERIALIZER_CLASS_CONFIG) == null)
         return newConfigs;
     }
 
-    boolean maybeOverrideEnableAutoCommit() {
-        Optional<String> groupId = Optional.ofNullable(getString(CommonClientConfigs.GROUP_ID_CONFIG));
-        boolean enableAutoCommit = getBoolean(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG);
-        if (!groupId.isPresent()) { // overwrite in case of default group id where the config is not explicitly provided
-            if (!originals().containsKey(ENABLE_AUTO_COMMIT_CONFIG)) {
-                enableAutoCommit = false;
-            } else if (enableAutoCommit) {
-                throw new InvalidConfigurationException(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG + " cannot be set to true when default group id (null) is used.");
-            }
-        }
-        return enableAutoCommit;
-    }
-
```
I also moved this to `ConsumerUtils` for the same reason as above.

kirktrue replied:
This is now outdated. It's back in `ConsumerConfig`.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -250,7 +255,7 @@ public PrototypeAsyncConsumer(final Time time,
             // no coordinator will be constructed for the default (null) group id
             if (!groupId.isPresent()) {
                 config.ignore(ConsumerConfig.AUTO_COMMIT_INTERVAL_MS_CONFIG);
-                //config.ignore(ConsumerConfig.THROW_ON_FETCH_STABLE_OFFSET_UNSUPPORTED);
+                config.ignore(ConsumerUtils.THROW_ON_FETCH_STABLE_OFFSET_UNSUPPORTED);
```
Moving the package-level variable and method from `ConsumerConfig` also fixes this issue.

kirktrue replied:
I've moved the core variable back to `ConsumerConfig` but left a copy in `ConsumerUtils` with an explanation of why it's redefined there.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -982,23 +987,6 @@ private boolean initWithCommittedOffsetsIfNeeded(Timer timer) {
         }
     }
 
-    // This is here temporary as we don't have public access to the ConsumerConfig in this module.
-    public static Map<String, Object> appendDeserializerToConfig(Map<String, Object> configs,
-                                                                 Deserializer<?> keyDeserializer,
-                                                                 Deserializer<?> valueDeserializer) {
-        // validate deserializer configuration, if the passed deserializer instance is null, the user must explicitly set a valid deserializer configuration value
-        Map<String, Object> newConfigs = new HashMap<>(configs);
-        if (keyDeserializer != null)
-            newConfigs.put(KEY_DESERIALIZER_CLASS_CONFIG, keyDeserializer.getClass());
-        else if (newConfigs.get(KEY_DESERIALIZER_CLASS_CONFIG) == null)
-            throw new ConfigException(KEY_DESERIALIZER_CLASS_CONFIG, null, "must be non-null.");
-        if (valueDeserializer != null)
-            newConfigs.put(VALUE_DESERIALIZER_CLASS_CONFIG, valueDeserializer.getClass());
-        else if (newConfigs.get(VALUE_DESERIALIZER_CLASS_CONFIG) == null)
-            throw new ConfigException(VALUE_DESERIALIZER_CLASS_CONFIG, null, "must be non-null.");
-        return newConfigs;
-    }
-
```
The original version of this method as it appeared in `ConsumerConfig` was moved to `ConsumerUtils` so now it can be used from here too 😄

philipnee replied:
Great find!

kirktrue replied:
This appears to be outdated. This is not in `ConsumerUtils` but stays in `ConsumerConfig`.

kirktrue committed:
Minor naming update

kirktrue committed:
Removed whitespace

kirktrue committed:
Reverted timeout in BaseAsyncConsumerTest back to original

kirktrue commented on core/src/test/scala/integration/kafka/api/IntegrationTestHarness.scala:
```diff
@@ -171,19 +170,6 @@ abstract class IntegrationTestHarness extends KafkaServerTestHarness {
     producer
   }
 
-  def createAsyncConsumer[K, V](keyDeserializer: Deserializer[K] = new ByteArrayDeserializer,
-                                valueDeserializer: Deserializer[V] = new ByteArrayDeserializer,
-                                configOverrides: Properties = new Properties,
-                                configsToRemove: List[String] = List()): PrototypeAsyncConsumer[K, V] = {
-    val props = new Properties
-    props ++= consumerConfig
-    props ++= configOverrides
-    configsToRemove.foreach(props.remove(_))
-    val consumer = new PrototypeAsyncConsumer[K, V](props, keyDeserializer, valueDeserializer)
-    consumers += consumer
-    consumer
-  }
-
```
This is no longer needed as we can create `KafkaConsumer` directly

kirktrue commented on tools/src/main/java/org/apache/kafka/tools/ClientCompatibilityTest.java:
```diff
@@ -430,8 +430,9 @@ public void testConsume(final long prodTimeMs) throws Throwable {
             // should work.
-            consumer.beginningOffsets(timestampsToSearch.keySet());
-            consumer.endOffsets(timestampsToSearch.keySet());
+            Map<TopicPartition, Long> beginningOffsets = consumer.beginningOffsets(timestampsToSearch.keySet());
+            Map<TopicPartition, Long> endingOffsets = consumer.endOffsets(timestampsToSearch.keySet());
+            log.trace("beginningOffsets: {}, endingOffsets: {}", beginningOffsets, endingOffsets);
```
This is super annoying.

I started getting errors from SpotBugs because the offsets methods were called but the return value was being ignored. This is a very brute force way of silencing the checker. I could not find a clean way to ignore the warning.

I also don't know why this is suddenly being caught by SpotBugs. From its perspective, nothing has changed in the `KafkaConsumer` API, right?

philipnee replied:
I find spotBugs flaky.  Sometimes it would complain about a bunch of stuff I never touched...

dajac replied:
This is weird.

dajac replied:
It would be great if we could revert this change.

kirktrue commented on tools/src/main/java/org/apache/kafka/tools/StreamsResetter.java:
```diff
@@ -281,7 +281,7 @@ private int maybeResetInputAndSeekToEndIntermediateTopicOffsets(final Map<Object
             if (!options.hasDryRun()) {
                 for (final TopicPartition p : partitions) {
-                    client.position(p);
+                    long pos = client.position(p);
```
This is another hack for bypassing the error from SpotBugs about ignored return values.

dajac replied:
It is strange that we have to change this now.

kirktrue replied:
I agree. I still plan to dig into it more.

kirktrue replied:
Reverted. I'll look for another solution.

kirktrue reviewed (commented):
Added some comments and origin story to the changes.

kirktrue committed:
Updated TaskAssignorIntegrationTest to handle the new delegate pattern

kirktrue committed:
Manually configuring the Consumer by specifying the group.protocol value

kirktrue commented:
@dajac This is ready for your review. The test failures are unrelated. Thanks!

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -670,168 +582,11 @@ public KafkaConsumer(Properties properties,
-    }
-
-    // visible for testing
-    KafkaConsumer(LogContext logContext,
```
This constructor is most likely used for injecting testing deps.  I wonder if we should just pass these components to the impl.

philipnee replied:
Giving it a second thought - I don't think we need it.  We just need to ensure the delegate is using the right class.

kirktrue replied:
Plus the testing dependencies differ wildly between the delegate implementations.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerDelegateCreator.java:
```diff
@@ -0,0 +1,132 @@
+    /**
+     * This is it! This is the core logic. It's extremely rudimentary.
+     */
+    private static boolean useNewConsumer(Map<?, ?> configs) {
```
The configuration is validated in the ConsumerConfig.java constructor; I wonder if we could directly invoke: `    KafkaConsumer(ConsumerConfig config, Deserializer<K> keyDeserializer, Deserializer<V> valueDeserializer) ` to avoid double validation.

kirktrue replied:
`useNewConsumer` was written to take a generic `Map` so that it can be called from all the constructors, including the ones that don't use `ConsumerConfig`.

dajac replied:
I made a similar comment earlier. Would it make sense to create `ConsumerConfig`, validate it and determine whether we should use the old or the new consumer?

kirktrue replied:
We could, yes. I will try to refactor to do that.

kirktrue replied:
I've refactored the code to use fewer constructors in the delegates and pass a fully-constructed `ConsumerConfig` to the delegates.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerDelegateCreator.java:
```diff
@@ -0,0 +1,132 @@
+        });
+    }
+
+    public <K, V> Consumer<K, V> create(Properties properties) {
```
could use `propsToMap(properties)` to invoke `public <K, V> Consumer<K, V> create(Map<String, Object> configs)`. 
`

kirktrue replied:
I wanted a specific constructor in `KafkaConsumer` to invoke the matching specific constructor of the delegate class. That way there's no weird manipulation/trickery in case the delegate constructors did things subtly different.

kirktrue replied:
I've refactored the code to use fewer constructors in the delegates and pass a fully-constructed `ConsumerConfig` to the delegates.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerDelegateCreator.java:
```diff
@@ -0,0 +1,132 @@
+                                        Deserializer<K> keyDeserializer,
+                                        Deserializer<V> valueDeserializer) {
+        return createInternal(() -> {
+            if (useNewConsumer(config.values())) {
```
There's a lot of duplicated code for
```
if (useNewConsumer(config.values())) {
                return new AsyncKafkaConsumer<>(config, keyDeserializer, valueDeserializer);
            } else {
                return new LegacyKafkaConsumer<>(config, keyDeserializer, valueDeserializer);
            }
```

kirktrue replied:
Yep. I don't really care for it, either. As above, I wanted to preserve a 1:1 mapping between the constructor that was invoked in `KafkaConsumer` and the constructor in the delegate. If we feel that's overkill, we can change it.

dajac replied:
I think that we should simplify the constructors in the delegates. It does not make sense to keep all of them.

kirktrue replied:
OK.

kirktrue replied:
I've refactored the code to use fewer constructors in the delegates and pass a fully-constructed `ConsumerConfig` to the delegates.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/LegacyKafkaConsumer.java:
```diff
@@ -0,0 +1,2616 @@
+import static org.apache.kafka.common.utils.Utils.swallow;
+
+/**
+ * A client that consumes records from a Kafka cluster.
```
Can we add a description that this module will only be used if the user decides to use the `generic`/JoinGroup/old protocol?

kirktrue replied:
Will do.

kirktrue replied:
Done. LMK if you'd like the wording changed.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -108,13 +105,13 @@
 /**
- * This prototype consumer uses an {@link ApplicationEventHandler event handler} to process
- * {@link ApplicationEvent application events} so that the network IO can be processed in a dedicated
+ * This {@link Consumer} implementation uses an {@link ApplicationEventHandler event handler} to process
```
Similarly - Can we add a description stating the module is used to support KIP-848 protocol?

kirktrue replied:
Will do.

kirktrue replied:
Done. If you want the verbiage changed in some fashion, just let me know. Thanks!

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerUtils.java:
```diff
@@ -211,4 +226,17 @@ else if (t instanceof KafkaException)
+    public static boolean maybeOverrideEnableAutoCommit(ConsumerConfig config) {
+        Optional<String> groupId = Optional.ofNullable(config.getString(CommonClientConfigs.GROUP_ID_CONFIG));
+        boolean enableAutoCommit = config.getBoolean(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG);
+        if (!groupId.isPresent()) { // overwrite in case of default group id where the config is not explicitly provided
```
Did we add a check to invalidate empty groupId?  https://issues.apache.org/jira/browse/KAFKA-14438

philipnee replied:
I was also wondering if this should should happen in the ConsumerConfig

kirktrue replied:
I didn't add it in this PR, no. The Jira refers to doing that in the "next major release," so that would imply we should hold off and handle it outside this PR.

dajac commented:
@kirktrue Noted. I will be away tomorrow but I can’t take a look at it when I come back if nobody gets to it first.

kirktrue committed:
Updated documentation to reflect intended use cases for the legacy and async consumers

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -278,22 +283,22 @@ public PrototypeAsyncConsumer(final Time time,
-                                  int defaultApiTimeoutMs,
-                                  List<ConsumerPartitionAssignor> assignors,
-                                  String groupId) {
+    public AsyncKafkaConsumer(LogContext logContext,
```
I don't remember if this constructor is used for testing, if it is can we change the scope to package-private?

dajac replied:
+1 for making it package private.

kirktrue replied:
Done.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -321,20 +320,6 @@ public class ConsumerConfig extends AbstractConfig {
-     * Note: this is an internal configuration and could be changed in the future in a backward incompatible way
-     *
-     */
-    static final String THROW_ON_FETCH_STABLE_OFFSET_UNSUPPORTED = "internal.throw.on.fetch.stable.offset.unsupported";
```
I am not fully convinced by moving this one to ConsumerUtils. It seems to me that we are putting it a bit "no where".  I also noticed that we have `LEAVE_GROUP_ON_CLOSE_CONFIG` which is also an internal one. Do you know why we don't have an issue with this one? In the end, I wonder if we should rather create a `InternalConsumerConfig` class for those internal configs.

kirktrue replied:
I have reverted this change. I have left `THROW_ON_FETCH_STABLE_OFFSET_UNSUPPORTED` in `ConsumerConfig` but I've added a string constant in `ConsumerUtils` just so it can be referenced in the `internals` package.

kirktrue replied:
@dajac—just following up on this point, specifically. Are you OK with this change? It's still awkward, I know, so if you have other ideas, let me know. Thanks!

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -108,13 +106,19 @@
+ * This consumer implements the new consumer group protocol and is intended to be the default in coming releases.
  */
-public class PrototypeAsyncConsumer<K, V> implements Consumer<K, V> {
+public class AsyncKafkaConsumer<K, V> implements Consumer<K, V> {
```
I wonder if we should make it package private. Thoughts?

kirktrue replied:
Done.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -148,30 +152,38 @@ public class PrototypeAsyncConsumer<K, V> implements Consumer<K, V> {
-    public PrototypeAsyncConsumer(final Properties properties,
-                                  final Deserializer<K> keyDeserializer,
-                                  final Deserializer<V> valueDeserializer) {
+    public AsyncKafkaConsumer(Map<String, Object> configs) {
```
I really wonder if we need all those constructors now that we have the delegate creator. For instance, would it be possible to only keep the following ones and keep all the conversion to `ConsumerConfig` in the delegate creator?
```
public AsyncKafkaConsumer(final ConsumerConfig config,
                              final Deserializer<K> keyDeserializer,
                              final Deserializer<V> valueDeserializer) {
        this(Time.SYSTEM, config, keyDeserializer, valueDeserializer);
    }

    public AsyncKafkaConsumer(final Time time,
                              final ConsumerConfig config,
                              final Deserializer<K> keyDeserializer,
                              final Deserializer<V> valueDeserializer) {
```

kirktrue replied:
Yep-done.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerDelegateCreator.java:
```diff
@@ -0,0 +1,132 @@
+     * This is it! This is the core logic. It's extremely rudimentary.
+     */
+    private static boolean useNewConsumer(Map<?, ?> configs) {
+        Object groupProtocol = configs.get("group.protocol");
```
nit: Could we use `getString`?

kirktrue replied:
Done.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerDelegateCreator.java:
```diff
@@ -0,0 +1,132 @@
+    private static boolean useNewConsumer(Map<?, ?> configs) {
+        Object groupProtocol = configs.get("group.protocol");
+
+        // Takes care of both the null and type checks.
+        if (!(groupProtocol instanceof String))
+            return false;
```
Could it be `null` based on the config definition?

kirktrue replied:
This code was written before the `ConsumerConfig` changes were in. I've updated it to use the `ConsumerConfig` directly.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerDelegateCreator.java:
```diff
@@ -0,0 +1,132 @@
+        if (!(groupProtocol instanceof String))
+            return false;
+
+        return ((String) groupProtocol).equalsIgnoreCase("consumer");
```
We have an enum for this. It would be better to reuse it here.

kirktrue replied:
Agreed. This code was written before the `ConsumerConfig` changes were in. I've updated it to use the `ConsumerConfig` directly.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerUtils.java:
```diff
@@ -211,4 +226,17 @@ else if (t instanceof KafkaException)
         }
     }
+
+    public static boolean maybeOverrideEnableAutoCommit(ConsumerConfig config) {
```
nit: Not your fault but I find the name of the method a little weird because it does not override anything. It just returns whether auto commit should be enabled or not.

kirktrue replied:
It has now been changed to override the configuration. LMK if you think that is a good idea, or if I should revert it.

kirktrue replied:
@dajac—following up on this—are you OK with this change?

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/LegacyKafkaConsumer.java:
```diff
@@ -0,0 +1,2623 @@
+/*
```
For my understanding, this is basically a copy of the old KafkaConsumer without any changes, right?

kirktrue replied:
Yes. The only changes are the removal of the extra constructors.

dajac commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/LegacyKafkaConsumerTest.java:
```diff
@@ -164,7 +163,7 @@
  * the test.
  */
-public class KafkaConsumerTest {
+public class LegacyKafkaConsumerTest {
```
For my understanding, do we have similar tests for the new consumer? Or do we need to run a subset of those for it?

kirktrue replied:
We have similar tests. My other attempt at the delegate work was much _deeper_ so that we could reuse this test and run it against both implementations, but it got out of hand.

dajac replied:
Are you referring to `AsyncKafkaConsumerTest`? I wonder if we should try a bit more to keep this one for both implementation. In this PR, we could perhaps keep it as `KafkaConsumerTest` and configure it to only run against the legacy consumer and we could parameterised it further as a follow-up. Without `KafkaConsumerTest`, we don't have any other tests validating the main "interface" and there is a ton of tests in this one.

kirktrue replied:
I have updated all of the tests to support parameterization. However, about half of them only support the legacy protocol because when we exercise the new protocol, the test fails. There is an explanation in the Javadoc of the method that is used as a 'source' for the parameters that elaborates.

dajac commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/LegacyKafkaConsumerTest.java:
```diff
@@ -2582,7 +2581,7 @@ private FetchResponse fetchResponse(TopicPartition partition, long fetchOffset,
 
-    private KafkaConsumer<String, String> newConsumer(Time time,
+    private LegacyKafkaConsumer<String, String> newConsumer(Time time,
                                                       KafkaClient client,
```
nit: The indentation of the arguments are not incorrect in a few methods.

kirktrue replied:
Thanks. Fixed.

dajac reviewed (commented):
@kirktrue Thanks for the PR. I just made a first pass on it and I left a few comments for consideration.

kirktrue committed:
Merge branch 'trunk' into KAFKA-15277-consumer-delegates-take-2

kirktrue committed:
Fixes to constructors, constructor visibility, configuration, and comments

kirktrue committed:
Fixed LegacyKafkaConsumerTest indentation issues

kirktrue committed:
Restored "internal.throw.on.fetch.stable.offset.unsupported" and maybeOverrideEnableAutoCommit to ConsumerConfig

kirktrue committed:
Found two missing uses of "internal.throw.on.fetch.stable.offset.unsupported"

kirktrue committed:
Minor refactoring for maybeOverrideEnableAutoCommit

kirktrue committed:
Added THROW_ON_FETCH_STABLE_OFFSET_UNSUPPORTED constant back to ConsumerUtils

kirktrue committed:
Changed ConsumerDelegateCreator to use the GroupProtocol enum

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/LegacyKafkaConsumer.java:
```diff
@@ -0,0 +1,2554 @@
+ * when invoking an API that is not available on the running broker version.
+ * <p>
+ *
+ * <h3>Offsets and Consumer Position</h3>
```
A lot of this massive comment really applies to `KafkaConsumer`.

kirktrue replied:
Agreed. I had a comment elsewhere where I don't really like leaving it here, but removing it also seems odd.

kirktrue replied:
Updated.

dajac replied:
I wonder if we should just remove all the main/common javadoc from this class and keep it in KafkaConsumer. The issue that I see is that this one will get out of sync. What do you think?

kirktrue replied:
Ideally we'd keep any implementation-specific comments in `LegacyKafkaConsumer` and remove them from `KafkaConsumer`. Would you be OK to do that as a separate PR?

kirktrue replied:
There wasn't much documentation that was `LegacyKafkaConsumer`-specific, so I gutted ~99% of the comments.

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerConfigTest.java:
```diff
@@ -67,17 +67,11 @@ public void testOverrideClientId() {
         ConsumerConfig config = new ConsumerConfig(properties);
-        boolean overrideEnableAutoCommit = config.maybeOverrideEnableAutoCommit();
-        assertFalse(overrideEnableAutoCommit);
+//        boolean overrideEnableAutoCommit = InternalConsumerConfig.maybeOverrideEnableAutoCommit(config);
```
This comment strikes me as strange.

kirktrue replied:
Agreed. I revised this test method.

kirktrue committed:
Removed unnecessary import statement

kirktrue committed:
Using ConsumerConfig and GroupProtocol to configure Consumer in BaseAsyncConsumerTest

kirktrue committed:
Update enable.auto.commit override test case

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -695,17 +696,16 @@ else if (newConfigs.get(VALUE_DESERIALIZER_CLASS_CONFIG) == null)
+    private void maybeOverrideEnableAutoCommit(Map<String, Object> configs) {
         Optional<String> groupId = Optional.ofNullable(getString(CommonClientConfigs.GROUP_ID_CONFIG));
-        boolean enableAutoCommit = getBoolean(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG);
+        boolean enableAutoCommit = values().containsKey(ENABLE_AUTO_COMMIT_CONFIG) ? getBoolean(ENABLE_AUTO_COMMIT_CONFIG) : false;
```
For my understanding, why do we need to check `values().containsKey(ENABLE_AUTO_COMMIT_CONFIG)`?

kirktrue replied:
We can't call `getBoolean` because if the value isn't there, it throws an exception. We have to look at the `originals` to get the values that were passed in by the user.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -278,22 +260,22 @@ public PrototypeAsyncConsumer(final Time time,
-                                  int defaultApiTimeoutMs,
-                                  List<ConsumerPartitionAssignor> assignors,
-                                  String groupId) {
+    public AsyncKafkaConsumer(LogContext logContext,
```
Do we need to keep this one public?

kirktrue replied:
No. Made package private.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerDelegateCreator.java:
```diff
@@ -0,0 +1,65 @@
+    public <K, V> Consumer<K, V> create(ConsumerConfig config,
+                                        Deserializer<K> keyDeserializer,
+                                        Deserializer<V> valueDeserializer) {
+
```
nit: We could remove this empty line.

kirktrue replied:
Done.

dajac commented on clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerConfigTest.java:
```diff
@@ -66,18 +66,24 @@ public void testOverrideClientId() {
+
+        // Verify that explicitly disabling 'enable.auto.commit' still works.
+        properties.setProperty(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, Boolean.FALSE.toString());
+        assertDoesNotThrow(() -> new ConsumerConfig(properties));
```
nit: Would it be better to actually assert the value of the property after the ConsumerConfig is created?

kirktrue replied:
Done.

dajac commented on clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerConfigTest.java:
```diff
@@ -66,18 +66,24 @@ public void testOverrideClientId() {
+
+        // Verify that then adding 'group.id' to the mix allows it to pass OK.
+        properties.setProperty(ConsumerConfig.GROUP_ID_CONFIG, "test-group");
+        assertDoesNotThrow(() -> new ConsumerConfig(properties));
```
nit: ditto.

kirktrue replied:
Done.

dajac reviewed (commented):
@kirktrue Thanks for the update. Overall, it looks pretty good. I left a few comments for consideration.

Also, it seems that we don't have a test which verifies that the correct "internal" consumer is created based on the config. Is there a way to verify this?

kirktrue committed:
Removed dead store

kirktrue committed:
Reverted change to ClientCompatibilityTest

kirktrue committed:
Made AsyncKafkaConsumer constructor package private

kirktrue committed:
Removed superfluous newline in ConsumerDelegateCreator

kirktrue committed:
Reworked maybeOverrideEnableAutoCommit again

kirktrue committed:
Updates to reuse KafkaConsumerTest with both delegate implementations

kirktrue committed:
Added "Visible for testing" comments to AsyncKafkaConsumer's testing constructors

kirktrue committed:
Waiting for ConsumerNetworkThread to complete close() before returning

kirktrue committed:
Added a comment to explain why some tests are "genericGroupProtocolOnly"

kirktrue committed:
Update ConsumerConfigTest to validate the 'enable.auto.commit' configuration is expected

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerDelegate.java:
```diff
@@ -0,0 +1,36 @@
+
+/**
+ * This extension interface provides a handful of methods to expose internals of the {@link Consumer} for
+ * various tests.
```
nit: Should we make it clear that this interface is only used internally?

kirktrue replied:
Will do.

kirktrue replied:
Done.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerDelegate.java:
```diff
@@ -0,0 +1,36 @@
+ */
+public interface ConsumerDelegate<K, V> extends Consumer<K, V> {
+
+    String getClientId();
```
nit: I understand that you reused the existing name here. However, we usually don't prefix getters with `get`. Should we fix it while here?

kirktrue replied:
Good idea.

kirktrue replied:
Done.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerDelegate.java:
```diff
@@ -0,0 +1,36 @@
+
+    String getClientId();
+
+    Metrics metricsInternal();
```
This one was not there before. What's the reason for introducing it? I wonder if we should just keeping using `Consumer.metrics()`.

kirktrue replied:
There's already an existing `metrics()` method in `Consumer`:

```
Map<MetricName, ? extends Metric> metrics();
```

I will take another look to see if the uses of `metricsInternal()` can be reworked to use `metrics()` and then the whole thing can be avoided.

dajac replied:
Thanks.

kirktrue replied:
It doesn't seem like this will be trivial to change the references to `Metrics` entirely. There are a couple of cases, but several tests use the 'metrics reporters' that the `Metrics` class exposes. As such, we would need to keep something like `metricsInternal()` around.

Would it help to change the method name from `metricsInternal()` to something else?

dajac replied:
Ack. Let's keep it then. How about `metricsRegistry()`? I leave the name up to you. I don't have a strong opinion on it.

kirktrue replied:
Renamed.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerDelegateCreator.java:
```diff
@@ -0,0 +1,114 @@
+ * <p/>
+ *
+ * The current logic for the {@code ConsumerCreator} inspects the incoming configuration and determines if
+ * it is using the new KIP-848 consumer protocol or if it should fall back to the existing, legacy group protocol.
```
nit: `the new KIP-848 consumer protocol` -> `the new consumer group protocol (KIP-848)`?

kirktrue replied:
Will change.

kirktrue replied:
Done.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java:
```diff
@@ -247,12 +247,10 @@ private void closeInternal(final Duration timeout) {
-            } catch (InterruptedException e) {
-                log.error("Interrupted while waiting for consumer network thread to complete", e);
-            }
+        try {
```
Why do we need this change?

kirktrue replied:
Because one of the unit tests fails without it 😄

In the existing `KafkaConsumer.close()`, the coordinator eventually calls this method:

```java
    private void closeHeartbeatThread() {
        HeartbeatThread thread;
        synchronized (this) {
            if (heartbeatThread == null)
                return;
            heartbeatThread.close();
            thread = heartbeatThread;
            heartbeatThread = null;
        }
        try {
            thread.join();
        } catch (InterruptedException e) {
            log.warn("Interrupted while waiting for consumer heartbeat thread to close");
            throw new InterruptException(e);
        }
    }
```

By blocking via `Thread.join()`, it allows the coordinator to close fully before its underlying resources (client, etc.) are closed. In contrast, our background thread was only waiting up to the timeout, and then proceeded with closing the other resources even though the background thread may not have completed its cleanup.

If you'd prefer, I can save this change for a separate PR?

dajac replied:
We can keep it here. I just wanted to understand the reason behind it.

dajac commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -222,42 +214,74 @@ public void cleanup() {
+    }
+
+    @ParameterizedTest
+    @MethodSource("bothGroupProtocols")
```
nit: The `EnumSource` is pretty handy for this: `@EnumSource(value = GroupProtocol.class, names = {"GENERIC", "CONSUMER"})`. Any reason not to use it?

kirktrue replied:
Because I didn't know of its existence 😆

I'll make that change.

kirktrue replied:
Done.

dajac commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -222,42 +214,74 @@ public void cleanup() {
+        return Arrays.stream(GroupProtocol.values()).map(Arguments::of).collect(Collectors.toList());
+    }
+
+    /**
+     * A given test may choose to use the {@link GroupProtocol#GENERIC generic group protocol} for a number of reasons.
+     * Among the reasons for a test to do so is because it...
+     *
+     * <ul>
+     *     <li>
+     *         ...exercises rebalancing logic that is not yet implemented in the
+     *         {@link GroupProtocol#CONSUMER consumer group protocol}.
+     *     </li>
+     *     <li>...includes topic metadata that is not yet implemented in the consumer group protocol.</li>
+     *     <li>...fails, possibly due to the omission of functionality in the consumer group protocol.</li>
+     *     <li>...uses logic, timing, etc. that are not applicable to the consumer group protocol.</li>
+     * </ul>
+     *
+     * Less than half of the tests for the consumer group protocol pass as of now, but it's very tedious to
+     * investigate at this point due to known bugs and missing functionality.
+     */
```
This is kind of obvious. Should we remove it? In my opinion, we should aim for making all of them pass -- not in this PR, obviously -- and if one cannot, we should add a comment explaining why.

kirktrue replied:
Just so I'm clear, you'd like to me to remove this method and then add a more specific reason for skipping the new consumer group protocol at each test method?

dajac replied:
Yeah, I would also use `EnumSource` for those ones and I would remove this comment. We should rather track the missing parts with Jiras.

kirktrue replied:
@dajac—I've made the change to use `EnumSource` everywhere. Thanks for that suggestion 😄

As far as the logistics for tracking these ~29 bugs via Jiras—do you suggest we file separate Jiras for each test failure, an "uber" Jira for all of the failures, or something else?

dajac replied:
@kirktrue It would be better to group them somehow but I let you judge.

kirktrue replied:
> It would be better to group them somehow

Sounds good. I'll start the investigation and open Jiras as I find new causes.

dajac commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -721,16 +802,17 @@ public void verifyNoCoordinatorLookupForManualAssignmentWithSeek() {
-    @Test
-    public void verifyNoCoordinatorLookupForManualAssignmentWithOffsetCommit() {
+    @ParameterizedTest
+    @MethodSource("genericGroupProtocolOnly")
```
Why is this one only applied to the generic protocol? It seems that it should also work for the new one.

kirktrue replied:
Yes, it should. I spent about half a day attempting to investigate the underlying issue. I wasn't sure if we should spend that time in this PR or in separate PRs. I sent a separate message to that end, and I'll circle back to these later.

kirktrue replied:
Resolving this conversation as it is being handled in a PR-level conversation.

dajac commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -757,8 +839,9 @@ public void verifyNoCoordinatorLookupForManualAssignmentWithOffsetCommit() {
-    @Test
-    public void testFetchProgressWithMissingPartitionPosition() {
+    @ParameterizedTest
+    @MethodSource("genericGroupProtocolOnly")
```
ditto.

kirktrue replied:
Resolving this conversation as it is being handled in a PR-level conversation.

dajac commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -814,16 +897,17 @@ private void initMetadata(MockClient mockClient, Map<String, Integer> partitionC
-    @Test
-    public void testMissingOffsetNoResetPolicy() {
+    @ParameterizedTest
+    @MethodSource("genericGroupProtocolOnly")
```
ditto.

kirktrue replied:
Resolving this conversation as it is being handled in a PR-level conversation.

dajac commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -835,16 +919,17 @@ public void testMissingOffsetNoResetPolicy() {
-    @Test
-    public void testResetToCommittedOffset() {
+    @ParameterizedTest
+    @MethodSource("genericGroupProtocolOnly")
```
Same. Actually, it seems to me that all the tests relying on `consumer.assign` should work, no?

kirktrue replied:
Resolving this conversation as it is being handled in a PR-level conversation.

dajac commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -857,16 +942,17 @@ public void testResetToCommittedOffset() {
-    @Test
-    public void testResetUsingAutoResetPolicy() {
+    @ParameterizedTest
+    @MethodSource("genericGroupProtocolOnly")
```
ditto.

kirktrue replied:
Resolving this conversation as it is being handled in a PR-level conversation.

dajac commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -881,24 +967,26 @@ public void testResetUsingAutoResetPolicy() {
-    @Test
-    public void testCommitsFetchedDuringAssign() {
+    @ParameterizedTest
+    @MethodSource("genericGroupProtocolOnly")
```
ditto.

kirktrue replied:
Resolving this conversation as it is being handled in a PR-level conversation.

dajac commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -1893,8 +2014,9 @@ private void consumerCloseTest(final long closeTimeoutMs,
-    public void testPartitionsForNonExistingTopic() {
+    @ParameterizedTest
+    @MethodSource("genericGroupProtocolOnly")
+    public void testPartitionsForNonExistingTopic(GroupProtocol groupProtocol) {
```
I suppose that we will update all those tests to support the new consumer at some point, right?

kirktrue replied:
Most, yes. There are some `enforceRebalance()` tests which—as I understand it—don't make sense to "port" over to the new group protocol.

dajac reviewed (commented):
@kirktrue Thanks for the update. The PR looks pretty good. I left a few comments for consideration.

Could you also elaborate a bit more on how you plan to fix all the tests in KafkaConsumerTest? It does not have to be done in this PR but I think that we should have a plan for it.

kirktrue committed:
Update to KafkaConsumerTest to use EnumSource for parameterizing tests

kirktrue committed:
Changed KafkaConsumer/ConsumerDelegate method getClientId() to clientId()

kirktrue committed:
Added documentation changes for the ConsumerDelegate and ConsumerDelegateCreator

kirktrue commented:
`KafkaConsumerTest` is a test suite that exercised the functionality of the "existing" `KafkaConsumer` with 100 tests. With this change, the suite was refactored to allow each of those tests to be parameterized, such that it now tests both `LegacyKafkaConsumer` and `AsyncKafkaConsumer`.

Ideally, those 100 tests would pass for both implementations, and we would have 200 passing tests 🎉

However, though the `LegacyKafkaConsumer` still passes all 100 tests in `KafkaConsumerTest`, the `AsyncKafkaConsumer` fails on 56 tests. The failures can be classified into the following reasons:

1. The test is only applicable for the old group protocol
2. The rebalancing logic is not yet implemented
3. The topic metadata logic is not yet implemented
4. The low-level CTR logic is incorrect

To the first point—some of the tests directly reference RPC calls from the old protocol and thus those tests won't pass without some effort in refactoring the tests. It may make sense to leave those alone versus bending things around to make them apply to the new protocol. As for the second and third points, at this point in development, we don't expect tests that exercise rebalance or topic metadata logic to work. For the fourth point, we are "gaining" new tests that stress the CTR layer in ways that we haven't tickled with our existing unit tests or integration tests.

Given this situation, I see two options forward:

1. Fix failing tests as part of this PR
2. Note which tests are failing, file Jiras, and fix them in future PRs

The first option allows us to merge in the underlying delegate change sooner, preventing it from blocking end-to-end testing. The second option is more holistic and doesn't attempt to shoehorn a half-baked test suite into `trunk`.

I take the view that we gained 44 additional passing tests with this change and have a followup task of gaining upward of 56 more soon.

@dajac @AndrewJSchofield @philipnee @lianetm—any opinions?

dajac commented:
Thanks for the explanation, @kirktrue. I agree with you that we should not block this PR on fixing all those tests. We can continue to fix them in subsequent PRs. For the tests only applicable to the old protocol (1), I agree that it does not make sense to touch them. However, we may need to add equivalent ones for the new protocol, do we?

dajac commented:
@kirktrue The build is red. Could you please take a look?

kirktrue committed:
Switched to EnumSource for all test methods

39 tests use both group protocols. 61 use only the GENERIC group protocol; greakdown:

29 known bugs with CONSUMER group protocol
24 reference unimplemented rebalance logic
2 reference unimplemented topic metadata logic
3 reference GENERIC-specific RPCs
3 reference enforceRebalance()

kirktrue commented:
An update on the `KafkaConsumerTest`—

Of the 100 tests, 39 successfully pass using both group protocols. 61 have failures when using the `CONSUMER` group protocol, so they have been hardcoded to only use `GENERIC` for now. Of those 61 tests:

- 29 trigger bugs when using the new `AsyncKafkaConsumer` (AKA CTR consumer) that the `CONSUMER` group protocol leverages 
- 24 reference rebalance logic that is currently unimplemented in the `CONSUMER` group protocol
- 2 reference topic metadata logic that is currently unimplemented in the `CONSUMER` group protocol
- 3 reference RPCs that are used only by the `GENERIC` group protocol
- 3 reference the `enforceRebalance()` that the `CONSUMER` group protocol intentionally doesn't implement

It will be one of my top priorities to circle back to the 'bugs' to investigate and fix.

kirktrue commented:
> @kirktrue The build is red. Could you please take a look?

The build failures are due to the recent "Dodgy code Warnings" that SpotBugs is suddenly issuing.

SpotBugs reports three violations of the `RV_RETURN_VALUE_IGNORED_NO_SIDE_EFFECT` check because we're ignoring the return value of `Consumer` method calls in three instances.

* Two of the instances are in `ClientCompatibilityTest`'s `testConsume` method because it ignores the return value of `beginningOffsets()` and `endOffsets()`
* The third instances is in `StreamsResetter`'s `maybeResetInputAndSeekToEndIntermediateTopicOffsets` because it doesn't use the value that `position()` returns

I'm still in a quandary about why they're now showing up. I'll continue digging.

dajac commented:
> 3 reference the enforceRebalance() that the CONSUMER group protocol intentionally doesn't implement

Should we add a small comment to those to explain that this is not supported in the new protocol?

> I'm still in a quandary about why they're now showing up. I'll continue digging.

Thanks!

kirktrue commented:
> Should we add a small comment to those to explain that this is not supported in the new protocol?

There is a comment above the three tests that use `enforceRebalance()` that states:

```java
// NOTE: this test uses the enforceRebalance API which is not implemented in the CONSUMER group protocol.
```

Is that sufficient?

dajac commented:
> > Should we add a small comment to those to explain that this is not supported in the new protocol?
> 
> There is a comment above the three tests that use `enforceRebalance()` that states:
> 
> ```java
> // NOTE: this test uses the enforceRebalance API which is not implemented in the CONSUMER group protocol.
> ```
> 
> Is that sufficient?

Yes.

kirktrue committed:
Updated AsyncKafkaConsumer initWithCommittedOffsetsIfNeeded for empty partitions

LegacyKafkaConsumer will return true immediately if the set of initializing partitions is empty. Changing this to support more tests in KafkaConsumerTest.

kirktrue committed:
Fix to ignore warnings when tools don't use the return value from Consumer calls

kirktrue committed:
Changed ConsumerDelegate metricsInternal() method name to metricsRegistry()

kirktrue committed:
Updated Javadoc for KafkaConsumer and LegacyKafkaConsumer

kirktrue committed:
Removed unused imports from LegacyKafkaConsumer that broke build

dajac reviewed (approved):
LGTM. Thanks, @kirktrue. I just re-triggered a build and I will merge if it passes.

kirktrue commented:
Re-triggered a build due to the abundance of seemingly unrelated test failures.

For example, the first line of this test in `ClusterConnectionStatesTest` is failing across all builds because `localhost` is resolving to more than one IP address:

```java
@Test
public void testSingleIP() throws UnknownHostException {
    assertEquals(1, ClientUtils.resolve("localhost", singleIPHostResolver).size());

    connectionStates.connecting(nodeId1, time.milliseconds(), "localhost");
    InetAddress currAddress = connectionStates.currentAddress(nodeId1);
    connectionStates.connecting(nodeId1, time.milliseconds(), "localhost");
    assertSame(currAddress, connectionStates.currentAddress(nodeId1));
}
```

dajac commented:
re-triggered a build as the last one did not look good.

kirktrue committed:
Merge branch 'trunk' into KAFKA-15277-consumer-delegates-take-2

kirktrue commented:
> re-triggered a build as the last one did not look good.

I only just now updated my branch to include the revert commits, so I wouldn't necessarily expect that test run to successfully pass. I have started yet another build with the previous reverts and will keep an eye on the results.

Thanks @dajac!

kirktrue commented:
@dajac—here are the unique test failures with notes:

- integration.kafka.server.FetchFromFollowerIntegrationTest: testRackAwareRangeAssignor(String).quorum=kraft
  - KAFKA-15020
- integration.kafka.server.FetchFromFollowerIntegrationTest: testRackAwareRangeAssignor(String).quorum=zk
  - KAFKA-15020, same as above
- kafka.api.ClientIdQuotaTest: testQuotaOverrideDelete(String).quorum=kraft
  - KAFKA-8107
  - **_Will investigate_**
- kafka.api.ClientIdQuotaTest: testThrottledProducerConsumer(String).quorum=zk
  - KAFKA-8108
  - **_Will investigate_**
- kafka.api.GroupEndToEndAuthorizationTest: testAuthentications(String).quorum=kraft
  - No Jira, but [it is flaky on `trunk` and PR builds](https://ge.apache.org/scans/tests?search.relativeStartTime=P28D&search.rootProjectNames=kafka&tests.container=kafka.api.GroupEndToEndAuthorizationTest&tests.test=testAuthentications(String)%5B2%5D)  
- kafka.api.TransactionsTest: testBumpTransactionalEpoch(String).quorum=kraft
  - KAFKA-15099
- org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest: testMultiWorkerRestartOnlyConnector
  - KAFKA-15675
- org.apache.kafka.connect.integration.StartAndStopLatchTest: shouldReturnFalseWhenAwaitingForStopToNeverComplete
  - **_Will investigate_**
- org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest: testReplication()
  - No Jira, but [it is flaky on `trunk` and PR builds](https://ge.apache.org/scans/tests?search.relativeStartTime=P28D&search.rootProjectNames=kafka&tests.container=org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest&tests.test=testReplication())
- org.apache.kafka.controller.QuorumControllerTest: testBalancePartitionLeaders()
  - KAFKA-15052, marked as fixed
- org.apache.kafka.controller.QuorumControllerTest: testFenceMultipleBrokers()
  - No Jira, but [it is flaky on `trunk` and PR builds](https://ge.apache.org/scans/tests?search.relativeStartTime=P28D&search.rootProjectNames=kafka&tests.container=org.apache.kafka.controller.QuorumControllerTest&tests.test=testFenceMultipleBrokers())
- org.apache.kafka.server.log.remote.metadata.storage.TopicBasedRemoteLogMetadataManagerTest: testRemoteLogSizeCalculationWithSegmentsOfDifferentEpochs()
  - No Jira, but [it is flaky on `trunk` and PR builds](https://ge.apache.org/scans/tests?search.relativeStartTime=P28D&search.rootProjectNames=kafka&tests.container=org.apache.kafka.server.log.remote.metadata.storage.TopicBasedRemoteLogMetadataManagerTest&tests.test=testRemoteLogSizeCalculationWithSegmentsOfDifferentEpochs())
- org.apache.kafka.streams.integration.EOSUncleanShutdownIntegrationTest: shouldWorkWithUncleanShutdownWipeOutStateStore[exactly_once_v2]
  - KAFKA-9868, marked as fixed
- org.apache.kafka.streams.integration.IQv2StoreIntegrationTest; verifyStore[cache=true, log=true, supplier=TIME_ROCKS_WINDOW, kind=DSL]
  - KAFKA-13714, marked as fixed
- org.apache.kafka.streams.integration.RestoreIntegrationTest: shouldInvokeUserDefinedGlobalStateRestoreListener()
  - KAFKA-15659, marked as fixed
- org.apache.kafka.tools.MetadataQuorumCommandTest: [1] Type=Raft-Combined, Name=testDescribeQuorumStatusSuccessful, MetadataVersion=3.7-IV1, Security=PLAINTEXT
  - KAFKA-15104
- org.apache.kafka.trogdor.coordinator.CoordinatorTest: testTaskRequestWithOldStartMsGetsUpdated()
  - KAFKA-8115

kirktrue commented:
Update on test failures:

- `kafka.api.ClientIdQuotaTest`
  - `testQuotaOverrideDelete`: ran 50 times locally with no errors
  - `testThrottledProducerConsumer`: ran 75 times locally with no errors
- `org.apache.kafka.connect.integration.StartAndStopLatchTest`
  - `shouldReturnFalseWhenAwaitingForStopToNeverComplete`: ran 400 times locally with no errors

Closing and re-opening to trigger build...

dajac commented:
I had a look at all the failed tests from the last build and they look all unrelated to me. Therefore, I am going to merge this PR to trunk. Thanks @kirktrue for the patch!
